storagenode/storagenodedb/pieceinfo.go

```diff
@@ -88,64 +88,67 @@ func (db *pieceinfo) Get(ctx context.Context, satelliteID storj.NodeID, pieceID
 func (db *pieceinfo) Delete(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) error {
 	defer db.locked()()
 
-	_, err := db.db.Exec(`DELETE FROM pieceinfo WHERE satellite_id = ? AND piece_id = ?`, satelliteID, pieceID)
+	_, err := db.db.ExecContext(ctx, db.Rebind(`
+		DELETE FROM pieceinfo 
+		WHERE satellite_id = ? 
+		  AND piece_id = ?
+	`), satelliteID, pieceID)
 
 	return ErrInfo.Wrap(err)
 }
 
-// SpaceUsed calculates disk space used by all pieces
-func (db *pieceinfo) SpaceUsed(ctx context.Context) (int64, error) {
+// DeleteFailed marks piece as a failed deletion.
+func (db *pieceinfo) DeleteFailed(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID, now time.Time) error {
 	defer db.locked()()
 
-	var sum *int64
-	err := db.db.QueryRow(`SELECT SUM(piece_size) FROM pieceinfo;`).Scan(&sum)
-	if err == sql.ErrNoRows || sum == nil {
-		return 0, nil
-	}
-	return *sum, err
+	_, err := db.db.ExecContext(ctx, db.Rebind(`
+		UPDATE pieceinfo 
+		SET deletion_failed_at = ?
+		WHERE satellite_id = ? 
+		  AND piece_id = ?
+	`), now, satelliteID, pieceID)
+
+	return ErrInfo.Wrap(err)
 }
 
-// GetExpired gets orders that are expired and were created before some time
-func (db *pieceinfo) GetExpired(ctx context.Context, expiredAt time.Time) (info []pieces.Info, err error) {
-	var getExpiredSQL = `SELECT satellite_id, piece_id, piece_size, piece_expiration, uplink_piece_hash, certificate.peer_identity
-		FROM pieceinfo
-		INNER JOIN certificate ON pieceinfo.uplink_cert_id = certificate.cert_id
-		WHERE piece_expiration < ? ORDER BY satellite_id `
+// GetExpired gets pieceinformation identites that are expired.
+func (db *pieceinfo) GetExpired(ctx context.Context, expiredAt time.Time, limit int64) (infos []pieces.ExpiredInfo, err error) {
+	defer db.locked()()
 
-	rows, err := db.db.QueryContext(ctx, db.Rebind(getExpiredSQL), expiredAt)
+	rows, err := db.db.QueryContext(ctx, db.Rebind(`
+		SELECT satellite_id, piece_id, piece_size
+		FROM pieceinfo
+		WHERE piece_expiration < ? AND ((deletion_failed_at IS NULL) OR deletion_failed_at <> ?)
```

Comment:
I don't understand the purpose of this part: `deletion_failed_at <> ?`. In reality, it should be always satisfied. It's very hard to pick the same timestamp.

Reply:
Ideally it should just be `WHERE piece_expiration < ? AND deletion_failed_at <> ?`... however there's some weirdness going on with nulls...

The Collect loop uses the same timestamp for all of the batches within a timeframe. It could use a threshold, but that would make that expression even more complicated. Also not sure what will happen with clock and timezone changes here otherwise.

Reply:
Oh, I think I got it. This would prevent `GetExpired` to return the same piece with "deletion failed flag" more than once in the same cycle.